### PR TITLE
Fix the nsupdate module's TTL change detection

### LINKED ISSUE #39465: nsupdate doesn't check for ttl changes
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
nsupdate

##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = /home/user/project/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Jan  5 2018, 10:41:29) [GCC 7.2.1 20171224]
```

##### CONFIGURATION
```
DEFAULT_HOST_LIST(/home/user/project/ansible.cfg) = [u'/home/user/project/inventory.yml']
HOST_KEY_CHECKING(/home/user/project/ansible.cfg) = False
RETRY_FILES_ENABLED(/home/user/project/ansible.cfg) = False
```

##### OS / ENVIRONMENT
Ansible running on Arch Linux, managed host is Debian Stretch.

##### SUMMARY
The module does not check for the ttl when it checks whether a change is required.

##### STEPS TO REPRODUCE
```yaml
- nsupdate:
    server: "{{ dns_server}}"
    ttl: "{{ item }}"
    zone: "example.com"
    record: "test"
    type: "a"
    value: "127.0.0.1"
  with_items:
    - 100
    - 200
```

##### EXPECTED RESULTS
The task should first create the record with 100s ttl, then change it to 200s ttl. This should at least cause 1 change on the first run, later runs should always cause 2 changes.

##### ACTUAL RESULTS
```
TASK [dnsrecords : Configure dns records] **********************************************************************************************
ok: [dnshost] => (item=100)
ok: [dnshost] => (item=200)

PLAY RECAP *****************************************************************************************************************************
dnshost                    : ok=2    changed=0    unreachable=0    failed=0   
```

### PR DESCRIPTION
**Pull request by andreaso**
##### SUMMARY

Ideally I would have liked to compare the TTL as part of the
prerequisite check. Sadly that isn't supported by the RFC 2136 update
protocol. Hence the additional query.

Resolves #39465

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
nsupdate

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (devel 32436ea9a5) last updated 2018/07/18 21:20:13 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/andreas/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/andreas/labs/ansible/devel/lib/ansible
  executable location = /home/andreas/labs/ansible/devel/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
```

### CODE REVIEW

**Comment by Nothing4You** [comment]
@andreaso I just tested this patch, it does verify the TTL correctly, updating when it differs.

**Comment by andreaso** [comment]
@Nothing4You: Thanks. Feel like reverifying that, now I have rebased this PR against the zone-detection fix?

**Comment by Nothing4You** [comment]
@andreaso same result as before, works fine for me.

**Comment by andreaso** [comment]
@Nothing4You: Thanks!

@nerzhul: How about you, are you ok with this PR?

File: lib/ansible/modules/net_tools/nsupdate.py [hunk 1]
```diff
@@ -326,12 +331,26 @@ def record_exists(self):
...
 
+    def ttl_changed(self):
+        query = dns.message.make_query(self.fqdn, self.module.params['type'])
+
+        try:
+            lookup = dns.query.tcp(query, self.module.params['server'], timeout=10, port=self.module.params['port'])
+        except (socket_error, dns.exception.Timeout) as e:
+            self.module.fail_json(msg='DNS server error: (%s): %s' % (e.__class__.__name__, to_native(e)))
+
+        current_ttl = lookup.answer[0].ttl
```
Comment (nerzhul) [inline lib/ansible/modules/net_tools/nsupdate.py:351]: what if empty result ? This should be tested
also we can have a different TTL per record value, not only per record if i'm correct

File: lib/ansible/modules/net_tools/nsupdate.py [hunk 2]
```diff
@@ -190,6 +190,11 @@ def __init__(self, module):
             if self.zone[-1] != '.':
                 self.zone += '.'
 
+        if module.params['record'][-1] != '.':
```
Comment (nerzhul) [inline lib/ansible/modules/net_tools/nsupdate.py:193]: seems unrelated to this PR

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/net_tools/nsupdate.py (hunk 1 above)
Comment (andreaso) [reply lib/ansible/modules/net_tools/nsupdate.py:351]: In which scenario would we get an empty result? Wouldn't we have taken a different code path in that scenario?

A different TTL per value? How would that even work DNS wise?

File: lib/ansible/modules/net_tools/nsupdate.py (hunk 2 above)
Comment (andreaso) [reply lib/ansible/modules/net_tools/nsupdate.py:193]: How do you suggest I do the TTL lookup without having a fqdn to use for the lookup?

File: lib/ansible/modules/net_tools/nsupdate.py (hunk 1 above)
Comment (Nothing4You) [reply lib/ansible/modules/net_tools/nsupdate.py:351]: I just tested that in my bind

`$ nsupdate`
```
> server localhost
> update add ttldemo.msg.lol. 300 TXT "TTL 300"
> update add ttldemo.msg.lol. 600 TXT "TTL 600"
> send
```
After syncing the changes to disk both entries in the zone file have a TTL of 600.

`$ dig txt ttldemo.msg.lol @127.0.0.1`
```
;; ANSWER SECTION:
ttldemo.msg.lol.	600	IN	TXT	"TTL 300"
ttldemo.msg.lol.	600	IN	TXT	"TTL 600"
```
I haven't checked RFCs but at least bind doesn't seem to support that.

Comment (andreaso) [reply lib/ansible/modules/net_tools/nsupdate.py:351]: https://tools.ietf.org/html/rfc2181#section-5.2

> ...
> Consequently the use of differing TTLs in an RRSet is hereby deprecated, the TTLs of all RRs in an RRSet must be the same.
> ...

File: lib/ansible/modules/net_tools/nsupdate.py (hunk 2 above)
Comment (nerzhul) [reply lib/ansible/modules/net_tools/nsupdate.py:193]: i see

File: lib/ansible/modules/net_tools/nsupdate.py (hunk 1 above)
Comment (nerzhul) [reply lib/ansible/modules/net_tools/nsupdate.py:351]: okay thanks for the clarification.

**Review by nerzhul — APPROVED** [review]

**Comment by nerzhul** [comment]
shipit

**Review by Nothing4You — APPROVED** [review]
